fake_quantize: respect device affinity

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39031 fake_quantize: respect device affinity**

Summary:

Makes the eager mode QAT prepare logic respect device affinity.
This fixes the issue where a module is on `cuda:0`, and running
the QAT prepare script would add observers on `cpu`.  Now it
will add them on the original device.

Test Plan:

```
python test/test_quantization.py TestDistributed.test_device_affinity
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21729272](https://our.internmc.facebook.com/intern/diff/D21729272)

vkuzo committed:
fake_quantize: respect device affinity

Summary:

Makes the eager mode QAT prepare logic respect device affinity.
This fixes the issue where a module is on `cuda:0`, and running
the QAT prepare script would add observers on `cpu`.  Now it
will add them on the original device.

Test Plan:

```
python test/test_quantization.py TestDistributed.test_device_affinity
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on test/quantization/test_workflow_module.py:
```diff
@@ -782,3 +782,36 @@ def forward(self, x):
+                return x
+
+        model = Model()
+        model.qconfig = torch.quantization.get_default_qat_qconfig('fbgemm')
```
Do we need to add the skipIfFBGEMM decorator as well?

supriyar reviewed (commented):
Looks good, left a comment

vkuzo committed:
Update on "fake_quantize: respect device affinity"


Summary:

Makes the eager mode QAT prepare logic respect device affinity.
This fixes the issue where a module is on `cuda:0`, and running
the QAT prepare script would add observers on `cpu`.  Now it
will add them on the original device.

Test Plan:

```
python test/test_quantization.py TestDistributed.test_device_affinity
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21729272](https://our.internmc.facebook.com/intern/diff/D21729272)

[ghstack-poisoned]

supriyar reviewed (approved):
lgtm, I think test failures need to be addressed before submitting though.

vkuzo committed:
Update on "fake_quantize: respect device affinity"


Summary:

Makes the eager mode QAT prepare logic respect device affinity.
This fixes the issue where a module is on `cuda:0`, and running
the QAT prepare script would add observers on `cpu`.  Now it
will add them on the original device.

Test Plan:

```
python test/test_quantization.py TestDistributed.test_device_affinity
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21729272](https://our.internmc.facebook.com/intern/diff/D21729272)

[ghstack-poisoned]
